DTML.EduBot/Dialogs/RootDialog.cs

```diff
@@ -3,76 +3,265 @@
     using System;
     using System.Collections.Generic;
     using System.Collections.ObjectModel;
+    using System.Globalization;
     using System.Linq;
     using System.Threading.Tasks;
     using Autofac;
     using Microsoft.Bot.Builder.Dialogs;
     using Microsoft.Bot.Connector;
+    using DTML.EduBot.Common;
+    using DTML.EduBot.UserData;
 
     [Serializable]
     public class RootDialog : IDialog<string>
     {
+        private readonly int MAX_ATTEMPT = 2;
+
+        private const string Yes = "Yes";
+        private const string No = "No";
+
+        private static readonly IEnumerable<string> YesNoChoices = new ReadOnlyCollection<string>
+            (new List<String> {
+                Yes,
+                No});
+
         private const string ChatWithBot = "Chat With Bot";
-        private const string StartTheLessonPlan = "Start The Lesson Plan";
+        private const string StartTheLessonPlan = "Start English Lesson Plan";
 
-        private static readonly IEnumerable<string> Choices = new ReadOnlyCollection<string>
+        private static readonly IEnumerable<string> DialogChoices = new ReadOnlyCollection<string>
             (new List<String> {
                 ChatWithBot,
                 StartTheLessonPlan});
 
+        private const string TooManyAttemptMessage = "Sorry, you have attempted too many times :(";
+
         public Task StartAsync(IDialogContext context)
         {
             context.Wait(this.MessageReceivedAsync);
             return Task.CompletedTask;
         }
 
-        private Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
+        private async Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> result)
         {
-            string friendlyUserName = context.Activity.From.Name;
+            var userText = (await result).Text;
             
+            string detectedLanguageIsoCode = await MessageTranslator.IdentifyLangAsync(userText);
+
+            using (var scope = WebApiApplication.FindContainer().BeginLifetimeScope())
+            {
+                UserData.UserData userData = new UserData.UserData();
+                userData.UserId = context.Activity.From.Id;
+                userData.NativeLanguageIsoCode = detectedLanguageIsoCode;
+
+                scope.Resolve<IUserDataRepository>().UpdateUserData(userData);
+
+                if (MessageTranslator.DEFAULT_LANGUAGE.Equals(detectedLanguageIsoCode))
+                {
+                    await context.PostAsync(BotPersonality.UserNameQuestion);
+
+                    // detected it's english language
+                    context.Wait(this.UserNameReceivedAsync);
+                }
+                else
```

Comment:
Looking throughout this change, I noticed that there are many if/else statements due to language choice as well as different method implementations depending on whether the user is using the default language or not. I realized that we would basically have to do this everywhere else throughout the code.

I think we can avoid this. One solution would be to always call the translate methods and have the translate methods just return the English strings immediately if the language is the default language. Another option that could even avoid those if checks would be to have two classes that implement an interface. We choose which to instantiate based on the language chosen. One class would simply return the strings and the other would do the translation